File: tif_customization/tif_customization/page/school_dashboard/school_dashboard.js
frappe.pages["school-dashboard"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "School Dashboard",
		single_column: true,
	});

	new frappe.tif_customization.SchoolDashboard(page).make();
};

frappe.tif_customization = frappe.tif_customization || {};

frappe.tif_customization.SchoolDashboard = class SchoolDashboard {
	constructor(page) {
		this.page = page;
		this.school_filter = null;
		this.school_name_filter = null;
		this.field_officer_name_filter = null;
		this.status_filter = null;
		this.tps_filter = null;
		this.qps_filter = null;
		this.cee_filter = null;
		this.workshop_status_filter = null;
		this.include_upcoming_trainings_filter = null;

		this.debounced_load = frappe.utils.debounce(() => this.load_data(), 300);
	}

	make() {
		this.make_filters();
		this.make_layout();
		this.bind_events();
		this.load_data();
	}

	make_filters() {
		this.school_filter = this.page.add_field({
			label: __("School"),
			fieldtype: "Link",
			fieldname: "school",
			options: "School",
			change: () => this.load_data(),
		});

		this.school_name_filter = this.page.add_field({
			label: __("School Name"),
			fieldtype: "Data",
			fieldname: "school_name",
			change: () => this.load_data(),
		});

		this.field_officer_name_filter = this.page.add_field({
			label: __("Field Officer Name"),
			fieldtype: "Data",
			fieldname: "field_officer_name",
			change: () => this.load_data(),
		});

		this.status_filter = this.page.add_field({
			label: __("Status"),
			fieldtype: "Select",
			fieldname: "status",
			options:
				"\nActive\nInactive\nClosed\nIn Process\nNot Interested\nDirect Requirement Received",
			change: () => this.load_data(),
		});

		this.tps_filter = this.page.add_field({
			label: __("TPS"),
			fieldtype: "Select",
			fieldname: "tps",
			options: "\nYes\nNo",
			change: () => this.load_data(),
		});

		this.qps_filter = this.page.add_field({
			label: __("QPS"),
			fieldtype: "Select",
			fieldname: "qps",
			options: "\nYes\nNo",
			change: () => this.load_data(),
		});

		this.cee_filter = this.page.add_field({
			label: __("CEE"),
			fieldtype: "Select",
			fieldname: "cee",
			options: "\nYes\nNo",
			change: () => this.load_data(),
		});

		this.workshop_status_filter = this.page.add_field({
			label: __("Workshop Status"),
			fieldtype: "Select",
			fieldname: "workshop_status",
			options: "\nIn Process\nNot Interested\nAgreed\nUnknown",
			change: () => this.load_data(),
		});

		this.include_upcoming_trainings_filter = this.page.add_field({
			label: __("Upcoming Trainings"),
			fieldtype: "Check",
			fieldname: "include_upcoming_trainings",
			change: () => this.load_data(),
		});

		this.page.set_primary_action(__("Refresh"), () => this.load_data(), "refresh");
	}

	make_layout() {
		const html = `
			<div class="school-dashboard-wrap" style="padding: 16px;">
				<div id="school-kpi-grid" class="row" style="margin-bottom: 16px;"></div>
				<div class="table-responsive" style="background: #fff; border: 1px solid #d1d8dd; border-radius: 8px;">
					<table class="table table-bordered" style="margin-bottom: 0;">
						<thead>
							<tr>
								<th>School Name</th>
								<th>Field Officer</th>
								<th>Status</th>
								<th>Services</th>
								<th>Remarks</th>
								<th>TPS</th>
								<th>QPS</th>
								<th>CEE</th>
								<th>Qaida/Guide Dispatch</th>
								<th>MQH Dispatch</th>
								<th>Workshop / Training</th>
								<th>Workshop Status</th>
							</tr>
						</thead>
						<tbody id="school-dashboard-tbody">
							<tr><td colspan="12" class="text-center text-muted">Loading...</td></tr>
						</tbody>
					</table>
				</div>
				<div id="upcoming-trainings-section" style="margin-top: 16px; display: none;"></div>
			</div>
		`;

		$(this.page.body).find(".school-dashboard-wrap").remove();
		$(this.page.body).append(html);
	}

	bind_events() {
		$(this.page.body).on("click", ".school-open-link", function () {
			const school = $(this).data("school");
			if (school) {
				frappe.set_route("Form", "School", school);
			}
		});

		$(this.page.body).on("click", ".school-remarks-link", (e) => {
			const school = $(e.currentTarget).data("school");
			const school_name = $(e.currentTarget).data("schoolName");
			if (school) this.open_remarks_dialog(school, school_name);
		});

		// Make text search feel like "search", not "filter-on-blur"
		if (this.school_name_filter && this.school_name_filter.$input) {
			this.school_name_filter.$input.on("input", () => this.debounced_load());
		}
		if (this.field_officer_name_filter && this.field_officer_name_filter.$input) {
			this.field_officer_name_filter.$input.on("input", () => this.debounced_load());
		}
	}

	load_data() {
		const school = this.school_filter ? this.school_filter.get_value() : "";
		const school_name = this.school_name_filter ? this.school_name_filter.get_value() : "";
		const field_officer_name = this.field_officer_name_filter
			? this.field_officer_name_filter.get_value()
			: "";
		const status = this.status_filter ? this.status_filter.get_value() : "";
		const tps = this.tps_filter ? this.tps_filter.get_value() : "";
		const qps = this.qps_filter ? this.qps_filter.get_value() : "";
		const cee = this.cee_filter ? this.cee_filter.get_value() : "";
		const workshop_status = this.workshop_status_filter
			? this.workshop_status_filter.get_value()
			: "";
		const include_upcoming_trainings = this.include_upcoming_trainings_filter
			? this.include_upcoming_trainings_filter.get_value()
			: 0;

		frappe.call({
			method: "tif_customization.tif_customization.page.school_dashboard.school_dashboard.get_dashboard_data",
			args: {
				school,
				school_name,
				field_officer_name,
				status,
				tps,
				qps,
				cee,
				workshop_status,
				include_upcoming_trainings,
			},
			callback: (r) => {
				const data = r.message || {};
				if (data.error) {
					frappe.msgprint(__("Unable to load dashboard: {0}", [data.error]));
					return;
				}
				this.render_kpis(data.kpis || {});
				this.render_rows(data.rows || []);
				this.render_upcoming_trainings(data.upcoming_trainings || [], include_upcoming_trainings);
			},
			error: () => {
				frappe.msgprint(__("Failed to load school dashboard data."));
			},
		});
	}

	render_kpis(kpis) {
		const card_data = [
			{ label: "Total Schools", value: kpis.total_schools || 0 },
			{ label: "Active Schools", value: kpis.active_schools || 0 },
			{ label: "Inactive Schools", value: kpis.inactive_schools || 0 },
			{ label: "Closed Schools", value: kpis.closed_schools || 0 },
			{ label: "In Process Schools", value: kpis.in_process_schools || 0 },
			{ label: "Not Interested", value: kpis.not_interested_schools || 0 },
			{
				label: "Direct Requirement Received",
				value: kpis.direct_requirement_received_schools || 0,
			},
			{ label: "TPS Schools", value: kpis.tps_schools || 0 },
			{ label: "QPS Schools", value: kpis.qps_schools || 0 },
			{ label: "CEE Schools", value: kpis.cee_schools || 0 },
			{ label: "Qaida/Guide Dispatch", value: this.format_number(kpis.qaida_guide_dispatch || 0) },
			{ label: "MQH Dispatch", value: this.format_number(kpis.mqh_dispatch || 0) },
			{ label: "Workshop In Process", value: kpis.workshop_in_process || 0 },
			{ label: "Workshop Not Interested", value: kpis.workshop_not_interested || 0 },
		];

		if (typeof kpis.upcoming_trainings !== "undefined") {
			card_data.unshift({ label: "Upcoming Trainings", value: kpis.upcoming_trainings || 0 });
		}

		const html = card_data
			.map(
				(card) => `
				<div class="col-md-4 col-sm-6" style="margin-bottom: 12px;">
					<div style="padding: 12px; border: 1px solid #d1d8dd; border-radius: 8px; background: #f8f9fa;">
						<div class="text-muted small">${frappe.utils.escape_html(card.label)}</div>
						<div style="font-size: 22px; font-weight: 600;">${frappe.utils.escape_html(String(card.value))}</div>
					</div>
				</div>
			`
			)
			.join("");

		$("#school-kpi-grid").html(html);
	}

	render_rows(rows) {
		if (!rows.length) {
			$("#school-dashboard-tbody").html(
				`<tr><td colspan="12" class="text-center text-muted">No data found.</td></tr>`
			);
			return;
		}

		const html = rows
			.map((row) => {
				const school_name = frappe.utils.escape_html(row.school_name || "");
				const workshop_training = frappe.utils.escape_html(row.workshop_training || "Not Set");
				const workshop_status = frappe.utils.escape_html(row.workshop_status || "Unknown");
				const qaida = this.format_number(row.qaida_guide_dispatch || 0);
				const mqh = this.format_number(row.mqh_dispatch || 0);
				const tps = this.as_yes_no(row.tps, "TPS is associated with this school");
				const qps = this.as_yes_no(row.qps, "QPS is associated with this school");
				const cee = this.as_yes_no(row.cee, "CEE is associated with this school");
				const status = frappe.utils.escape_html(row.status_display || row.status || "-");
				const services = frappe.utils.escape_html(row.services || "-");
				const field_officer =
					frappe.utils.escape_html(row.field_officer_name || "") ||
					frappe.utils.escape_html(row.field_officer || "-");
				const latest_remark = (row.latest_remark || "").trim();
				const remark_short = latest_remark
					? frappe.utils.escape_html(frappe.utils.ellipsis(latest_remark, 80))
					: "-";

				return `
					<tr>
						<td>
							<a href="javascript:void(0)" class="school-open-link" data-school="${frappe.utils.escape_html(row.school || "")}">
								${school_name || "-"}
							</a>
						</td>
						<td>${field_officer || "-"}</td>
						<td>${status}</td>
						<td>${services}</td>
						<td>
							<a href="javascript:void(0)" class="school-remarks-link" data-school="${frappe.utils.escape_html(row.school || "")}" data-school-name="${school_name}">
								${remark_short}
							</a>
						</td>
						<td>${tps}</td>
						<td>${qps}</td>
						<td>${cee}</td>
						<td class="text-right">${qaida}</td>
						<td class="text-right">${mqh}</td>
						<td>${workshop_training}</td>
						<td>${workshop_status}</td>
					</tr>
				`;
			})
			.join("");

		$("#school-dashboard-tbody").html(html);
	}

	render_upcoming_trainings(rows, include_upcoming_trainings) {
		const show = !!cint(include_upcoming_trainings);
		if (!show) {
			$("#upcoming-trainings-section").hide().empty();
			return;
		}

		const body = rows.length
			? `
				<div class="table-responsive" style="background: #fff; border: 1px solid #d1d8dd; border-radius: 8px;">
					<table class="table table-bordered" style="margin-bottom: 0;">
						<thead>
							<tr>
								<th>Date</th>
								<th>Session Category</th>
								<th>Trainer</th>
								<th>City</th>
								<th>Province</th>
								<th>Venue</th>
								<th class="text-right">Participants</th>
								<th class="text-right">Schools</th>
							</tr>
						</thead>
						<tbody>
							${rows
								.map((r) => {
									const date = frappe.utils.escape_html(r.training_date || "-");
									const category = frappe.utils.escape_html(r.training_session_category || "-");
									const trainer = frappe.utils.escape_html(r.training_trainer_name || "-");
									const city = frappe.utils.escape_html(r.training_city || "-");
									const province = frappe.utils.escape_html(r.training_province || "-");
									const venue = frappe.utils.escape_html(r.training_venue_name || "-");
									const participants = this.format_number(r.training_no_of_participants || 0);
									const schools = this.format_number(r.training_no_of_schools_attended || 0);
									return `
										<tr>
											<td>${date}</td>
											<td>${category}</td>
											<td>${trainer}</td>
											<td>${city}</td>
											<td>${province}</td>
											<td>${venue}</td>
											<td class="text-right">${participants}</td>
											<td class="text-right">${schools}</td>
										</tr>
									`;
								})
								.join("")}
						</tbody>
					</table>
				</div>
			`
			: `<div class="text-muted">No upcoming trainings found.</div>`;

		const html = `
			<div style="padding: 12px; border: 1px solid #d1d8dd; border-radius: 8px; background: #f8f9fa;">
				<div style="font-size: 16px; font-weight: 600; margin-bottom: 8px;">Upcoming Trainings</div>
				${body}
			</div>
		`;

		$("#upcoming-trainings-section").html(html).show();
	}

	open_remarks_dialog(school, school_name) {
		const dialog = new frappe.ui.Dialog({
			title: __("Remarks: {0}", [school_name || school]),
			fields: [
				{
					fieldname: "history_html",
					fieldtype: "HTML",
				},
				{
					fieldname: "new_remark",
					label: __("Add Remark"),
					fieldtype: "Small Text",
					reqd: 1,
				},
			],
			primary_action_label: __("Add"),
			primary_action: () => {
				const values = dialog.get_values();
				if (!values || !values.new_remark) return;
				frappe.call({
					method: "tif_customization.tif_customization.page.school_dashboard.school_dashboard.add_school_remark",
					args: { school, remark: values.new_remark },
					callback: () => {
						dialog.set_value("new_remark", "");
						this.refresh_remarks_history(dialog, school);
						this.load_data();
					},
				});
			},
		});

		dialog.show();
		this.refresh_remarks_history(dialog, school);
	}

	refresh_remarks_history(dialog, school) {
		const $wrapper = $(dialog.fields_dict.history_html.wrapper);
		$wrapper.html(`<div class="text-muted">Loading...</div>`);

		frappe.call({
			method: "tif_customization.tif_customization.page.school_dashboard.school_dashboard.get_school_remarks",
			args: { school },
			callback: (r) => {
				const rows = r.message || [];
				if (!rows.length) {
					$wrapper.html(`<div class="text-muted">No remarks yet.</div>`);
					return;
				}
				const html = rows
					.map((row) => {
						const by = frappe.utils.escape_html(row.owner || "");
						const when = row.creation ? frappe.datetime.str_to_user(row.creation) : "";
						const content = frappe.utils.escape_html(row.content || "");
						return `
							<div style="padding: 8px 0; border-bottom: 1px solid #eee;">
								<div class="text-muted small">${by} • ${frappe.utils.escape_html(when)}</div>
								<div>${content || "-"}</div>
							</div>
						`;
					})
					.join("");
				$wrapper.html(`<div style="max-height: 320px; overflow: auto;">${html}</div>`);
			},
			error: () => {
				$wrapper.html(`<div class="text-danger">Failed to load remarks.</div>`);
			},
		});
	}

	as_yes_no(value, expected) {
		return value === expected ? "Yes" : "No";
	}

	format_number(value) {
		return format_number(value, null, 0);
	}
}

File: tif_customization/tif_customization/page/school_dashboard/school_dashboard.py
import frappe
from frappe import _
from frappe.exceptions import PermissionError
from frappe.utils import cint, flt, nowdate


@frappe.whitelist()
def get_dashboard_data(
    school=None,
    school_name=None,
    status=None,
    field_officer_name=None,
    tps=None,
    qps=None,
    cee=None,
    workshop_status=None,
    include_upcoming_trainings=0,
):
    """Return KPI and school-wise data for School Dashboard."""
    try:
        include_upcoming_trainings = cint(include_upcoming_trainings)
        filters = {"docstatus": ["<", 2]}
        if school:
            filters["name"] = school
        if school_name:
            filters["school_name"] = ["like", f"%{school_name.strip()}%"]
        if status:
            normalized_status = _normalize_school_status(status)
            if normalized_status:
                filters["status"] = ["in", _status_filter_values(normalized_status)]
        if field_officer_name:
            matched_users = _get_users_matching(field_officer_name)
            if not matched_users:
                return {
                    "kpis": _empty_kpis(include_upcoming_trainings),
                    "rows": [],
                    "upcoming_trainings": _get_upcoming_trainings() if include_upcoming_trainings else [],
                }
            filters["field_officer"] = ["in", matched_users]
        if tps:
            filters["tps"] = (
                "TPS is associated with this school" if tps == "Yes" else ["!=", "TPS is associated with this school"]
            )
        if qps:
            filters["qps"] = (
                "QPS is associated with this school" if qps == "Yes" else ["!=", "QPS is associated with this school"]
            )
        if cee:
            filters["cee"] = (
                "CEE is associated with this school" if cee == "Yes" else ["!=", "CEE is associated with this school"]
            )

        rows = _get_school_rows(filters)
        if workshop_status:
            rows = [r for r in rows if (r.get("workshop_status") or "") == workshop_status]

        total_schools = len(rows)

        status_counts = {
            "Active": 0,
            "Inactive": 0,
            "Closed": 0,
            "In Process": 0,
            "Not Interested": 0,
            "Direct Requirement Received": 0,
            "Other": 0,
        }
        for r in rows:
            normalized = _normalize_school_status(r.get("status"))
            r["status_display"] = normalized
            if normalized in status_counts:
                status_counts[normalized] += 1
            else:
                status_counts["Other"] += 1

        active_schools = status_counts.get("Active", 0)
        tps_schools = sum(1 for r in rows if r.get("tps") == "TPS is associated with this school")
        qps_schools = sum(1 for r in rows if r.get("qps") == "QPS is associated with this school")
        cee_schools = sum(1 for r in rows if r.get("cee") == "CEE is associated with this school")

        qaida_guide_dispatch = sum(flt(r.get("qaida_guide_dispatch", 0)) for r in rows)
        mqh_dispatch = sum(flt(r.get("mqh_dispatch", 0)) for r in rows)
        in_process = sum(1 for r in rows if r.get("workshop_status") == "In Process")
        not_interested = sum(1 for r in rows if r.get("workshop_status") == "Not Interested")

        upcoming_trainings = []
        upcoming_trainings_count = None
        if include_upcoming_trainings:
            upcoming_trainings = _get_upcoming_trainings()
            upcoming_trainings_count = frappe.db.count(
                "Field Visit",
                {
                    "docstatus": ["<", 2],
                    "type": "Training",
                    "training_date": [">=", nowdate()],
                },
            )

        kpis = {
            "total_schools": total_schools,
            "active_schools": active_schools,
            "inactive_schools": status_counts.get("Inactive", 0),
            "closed_schools": status_counts.get("Closed", 0),
            "in_process_schools": status_counts.get("In Process", 0),
            "not_interested_schools": status_counts.get("Not Interested", 0),
            "direct_requirement_received_schools": status_counts.get("Direct Requirement Received", 0),
            "tps_schools": tps_schools,
            "qps_schools": qps_schools,
            "cee_schools": cee_schools,
            "qaida_guide_dispatch": qaida_guide_dispatch,
            "mqh_dispatch": mqh_dispatch,
            "workshop_in_process": in_process,
            "workshop_not_interested": not_interested,
        }
        if include_upcoming_trainings:
            kpis["upcoming_trainings"] = upcoming_trainings_count or 0

        return {"kpis": kpis, "rows": rows, "upcoming_trainings": upcoming_trainings}
    except Exception as exc:
        frappe.log_error(frappe.get_traceback(), "School Dashboard Error")
        return {"error": str(exc), "kpis": {}, "rows": []}


def _get_school_rows(filters):
    school_docs = frappe.get_all(
        "School",
        fields=[
            "name",
            "school_name",
            "status",
            "remarks",
            "field_officer",
            "tps",
            "qps",
            "cee",
            "books",
            "trainings",
            "training_type_allowed",
        ],
        filters=filters,
        limit_page_length=0,
    )

    if not school_docs:
        return []

    school_docnames = [d.get("name") for d in school_docs if d.get("name")]
    school_names = [d.get("school_name") for d in school_docs if d.get("school_name")]
    dispatch_map = _get_dispatch_map(school_names)
    visit_status_map = _get_latest_marketing_status_map(school_names)
    latest_remark_map = _get_latest_remark_map(school_docnames)
    field_officer_name_map = _get_user_full_name_map(
        list({d.get("field_officer") for d in school_docs if d.get("field_officer")})
    )

    rows = []
    for doc in school_docs:
        school_name = doc.get("school_name")
        dispatch = dispatch_map.get(school_name, {})
        visit_status = visit_status_map.get(school_name, "")
        workshop_status = _map_workshop_status(visit_status)
        workshop_training = (
            doc.get("trainings")
            or doc.get("training_type_allowed")
            or _("Not Set")
        )
        latest_remark = latest_remark_map.get(doc.get("name")) or doc.get("remarks")

        services = "-"
        if _normalize_school_status(doc.get("status")) == "Active":
            service_bits = []
            if (doc.get("books") or "").strip():
                service_bits.append("Books")
            if (doc.get("trainings") or "").strip() or (doc.get("training_type_allowed") or "").strip():
                service_bits.append("Trainings")
            if workshop_status in {"Agreed", "In Process"}:
                service_bits.append("Workshops")
            if service_bits:
                services = ", ".join(service_bits)

        rows.append(
            {
                "school": doc.get("name"),
                "school_name": school_name,
                "status": doc.get("status"),
                "latest_remark": latest_remark,
                "field_officer": doc.get("field_officer"),
                "field_officer_name": field_officer_name_map.get(doc.get("field_officer")) or "",
                "tps": doc.get("tps"),
                "qps": doc.get("qps"),
                "cee": doc.get("cee"),
                "qaida_guide_dispatch": flt(dispatch.get("qaida_guide_dispatch", 0)),
                "mqh_dispatch": flt(dispatch.get("mqh_dispatch", 0)),
                "workshop_training": workshop_training,
                "workshop_status": workshop_status,
                "services": services,
            }
        )

    return rows


def _get_dispatch_map(school_names):
    if not school_names:
        return {}

    rows = frappe.db.sql(
        """
        SELECT
            dn.customer_name AS school_name,
            SUM(
                CASE
                    WHEN UPPER(COALESCE(dni.item_name, '')) LIKE '%%QAIDA%%'
                        OR UPPER(COALESCE(dni.item_code, '')) LIKE '%%QAIDA%%'
                        OR UPPER(COALESCE(dni.item_name, '')) LIKE '%%GUIDE%%'
                        OR UPPER(COALESCE(dni.item_code, '')) LIKE '%%GUIDE%%'
                    THEN COALESCE(dni.qty, 0)
                    ELSE 0
                END
            ) AS qaida_guide_dispatch,
            SUM(
                CASE
                    WHEN UPPER(COALESCE(dni.item_name, '')) LIKE '%%MQH%%'
                        OR UPPER(COALESCE(dni.item_code, '')) LIKE '%%MQH%%'
                    THEN COALESCE(dni.qty, 0)
                    ELSE 0
                END
            ) AS mqh_dispatch
        FROM `tabDelivery Note` dn
        INNER JOIN `tabDelivery Note Item` dni ON dni.parent = dn.name
        WHERE dn.docstatus = 1
          AND IFNULL(dn.is_return, 0) = 0
          AND dn.customer_name IN %(school_names)s
        GROUP BY dn.customer_name
        """,
        {"school_names": tuple(school_names)},
        as_dict=True,
    )

    return {r.get("school_name"): r for r in rows}


def _get_latest_marketing_status_map(school_names):
    if not school_names:
        return {}

    rows = frappe.db.sql(
        """
        SELECT fv.school_name, fv.status
        FROM `tabField Visit` fv
        INNER JOIN (
            SELECT school_name, MAX(modified) AS last_modified
            FROM `tabField Visit`
            WHERE docstatus < 2
              AND type = 'Marketing'
              AND IFNULL(school_name, '') != ''
              AND school_name IN %(school_names)s
            GROUP BY school_name
        ) latest
            ON latest.school_name = fv.school_name
            AND latest.last_modified = fv.modified
        WHERE fv.docstatus < 2
          AND fv.type = 'Marketing'
        """,
        {"school_names": tuple(school_names)},
        as_dict=True,
    )
    return {r.get("school_name"): r.get("status") for r in rows}


def _map_workshop_status(marketing_status):
    status = (marketing_status or "").strip().lower()
    if status == "pending":
        return "In Process"
    if status == "not agreed":
        return "Not Interested"
    if status == "agreed":
        return "Agreed"
    return "Unknown"


def _is_active_status(status):
    return (status or "").strip().upper() == "ACTIVE"


def _title_case_status(status):
    if status == "NOT INTERESTED":
        return "Not Interested"
    return status.title()


def _normalize_school_status(status):
    value = (status or "").strip()
    if not value:
        return ""
    upper = value.upper()
    if upper in {"PENDING", "IN PROCESS"}:
        return "In Process"
    if upper in {"NOT INTERESTED", "NOT-INTERESTED"}:
        return "Not Interested"
    if upper in {"DIRECT REQUIREMENT RECEIVED", "DIRECT_REQUIREMENT_RECEIVED"}:
        return "Direct Requirement Received"
    if upper == "ACTIVE":
        return "Active"
    if upper == "INACTIVE":
        return "Inactive"
    if upper == "CLOSED":
        return "Closed"
    return value


def _status_filter_values(normalized_status):
    # Keep backward-compatibility with older values stored in DB (e.g. "Pending")
    if normalized_status == "In Process":
        return ["In Process", "Pending"]
    return [normalized_status]


def _get_users_matching(field_officer_name):
    term = (field_officer_name or "").strip()
    if not term:
        return []
    term_like = f"%{term}%"
    users = frappe.get_all(
        "User",
        filters={
            "enabled": 1,
            "name": ["not in", ["Administrator", "Guest"]],
            "full_name": ["like", term_like],
        },
        pluck="name",
        limit_page_length=200,
    )
    if users:
        return users

    # fallback: match by user id/email-like value
    return frappe.get_all(
        "User",
        filters={
            "enabled": 1,
            "name": ["not in", ["Administrator", "Guest"]],
            "name": ["like", term_like],
        },
        pluck="name",
        limit_page_length=200,
    )


def _get_user_full_name_map(user_ids):
    user_ids = [u for u in (user_ids or []) if u]
    if not user_ids:
        return {}
    rows = frappe.get_all("User", filters={"name": ["in", user_ids]}, fields=["name", "full_name"], limit_page_length=0)
    return {r.get("name"): (r.get("full_name") or "") for r in rows}


def _get_latest_remark_map(school_docnames):
    school_docnames = [s for s in (school_docnames or []) if s]
    if not school_docnames:
        return {}

    rows = frappe.db.sql(
        """
        SELECT reference_name AS school, content, creation
        FROM `tabComment`
        WHERE comment_type = 'Comment'
          AND reference_doctype = 'School'
          AND reference_name IN %(schools)s
        ORDER BY creation DESC
        """,
        {"schools": tuple(school_docnames)},
        as_dict=True,
    )

    latest = {}
    for r in rows:
        school = r.get("school")
        if school and school not in latest:
            latest[school] = frappe.utils.strip_html(r.get("content") or "").strip()
    return latest


def _get_upcoming_trainings():
    return frappe.get_all(
        "Field Visit",
        filters={
            "docstatus": ["<", 2],
            "type": "Training",
            "training_date": [">=", nowdate()],
        },
        fields=[
            "name",
            "training_date",
            "training_city",
            "training_province",
            "training_venue_name",
            "training_session_category",
            "training_trainer_name",
            "training_no_of_participants",
            "training_no_of_schools_attended",
        ],
        order_by="training_date asc, modified desc",
        limit_page_length=50,
    )


def _empty_kpis(include_upcoming_trainings):
    base = {
        "total_schools": 0,
        "active_schools": 0,
        "inactive_schools": 0,
        "closed_schools": 0,
        "in_process_schools": 0,
        "not_interested_schools": 0,
        "direct_requirement_received_schools": 0,
        "tps_schools": 0,
        "qps_schools": 0,
        "cee_schools": 0,
        "qaida_guide_dispatch": 0,
        "mqh_dispatch": 0,
        "workshop_in_process": 0,
        "workshop_not_interested": 0,
        "upcoming_trainings": 0,
    }
    if not cint(include_upcoming_trainings):
        base.pop("upcoming_trainings", None)
    return base


@frappe.whitelist()
def get_school_remarks(school):
    school = (school or "").strip()
    if not school:
        return []
    if not frappe.has_permission("School", ptype="read", doc=school):
        frappe.throw(_("Not permitted"), PermissionError)

    rows = frappe.db.sql(
        """
        SELECT name, owner, creation, content
        FROM `tabComment`
        WHERE comment_type = 'Comment'
          AND reference_doctype = 'School'
          AND reference_name = %(school)s
        ORDER BY creation DESC
        LIMIT 200
        """,
        {"school": school},
        as_dict=True,
    )
    for r in rows:
        r["content"] = frappe.utils.strip_html(r.get("content") or "").strip()
    return rows


@frappe.whitelist()
def add_school_remark(school, remark):
    school = (school or "").strip()
    remark = (remark or "").strip()
    if not school:
        frappe.throw(_("School is required"))
    if not remark:
        frappe.throw(_("Remark is required"))
    if not frappe.has_permission("School", ptype="read", doc=school):
        frappe.throw(_("Not permitted"), PermissionError)

    doc = frappe.get_doc("School", school)
    doc.add_comment("Comment", text=remark)
    return {"ok": True}
